### ISSUE TITLE
New Octane Rule: `require-tagless-components`

### ISSUE THREAD

**Issue by patocallaghan** [ember-cli/eslint-plugin-ember#541]
Pulled from https://github.com/ember-cli/eslint-plugin-ember/issues/533 which proposes a new set of rules for Octane related best practices

Only allow `@glimmer/component` base class, or tagName: '' if `@ember/component` is used

```js
// ### Good
import Component from '@glimmer/component'
export default class MyComponent extends Component {
  ...
}

import Component from '@ember/component'
export default class MyComponent extends Component {
  tagName = '';
}

import Component from '@ember/component'
export default Component.extend({
  tagName: ''
});


// ### Bad
// implicit `<div>`
import Component from '@ember/component'
export default class MyComponent extends Component {
  ...
}

import Component from '@ember/component'
export default Component.extend({
  ...
});

// custom tag
import Component from '@ember/component'
export default class MyComponent extends Component {
  tagName = 'span';
}

import Component from '@ember/component'
export default Component.extend({
  tagName: 'span'
});
```

If anyone would like to implement this please comment that you'll take it and work away.

**Comment by patocallaghan** [comment]
@rwjblue @Turbo87 could you sanity check the good/bad examples please?

**Comment by scalvert** [comment]
They make sense to me.

**Comment by bmish** [comment]
Probably want to call this `require-tagless-components` for consistency with other rule names.

**Comment by alexlafroscia** [comment]
I have a bit of experience with writing ESLint rules and would be interested in taking this one on!

**Comment by alexlafroscia** [comment]
Would there be a reason to detect usage of the `@tagName()` decorator that used to be part of `ember-decorators`? Or is that not necessary?

**Comment by patocallaghan** [comment]
@alexlafroscia I don't use ember-decorators so I didn't include it in my examples. If that's a common enough way of using Classic components then I guess we should?

**Comment by alexlafroscia** [comment]
I'm not sure how common it is these days -- there was a time when `tagName` as a class property was not supported and it was the _only_ way to do things, but I don't know how much that's still used these days.

**Comment by patocallaghan** [comment]
@rwjblue @scalvert any thoughts here re:ember-decorators support?


@alexlafroscia If it's a lot of work we could possibly focus on the common use cases for now and add `@tagName` support later?

**Comment by alexlafroscia** [comment]
I decided to leave it out -- I'm not sure it's really worth supporting, honestly. If people report getting tripped up by the lack of catching `@tagName` it can always be added later.

I called out the fact that it isn't supported as a caveat in the documentation that i added for the rule!

**Comment by scottkidder** [comment]
I'm in favor of supporting `@tagName` mostly because the [native class codemod](https://github.com/ember-codemods/ember-native-class-codemod) uses it when converting classic to native.

**Comment by Turbo87** [comment]
> I'm in favor of supporting `@tagName` mostly because the [native class codemod](https://github.com/ember-codemods/ember-native-class-codemod) uses it when converting classic to native.

IMHO we should change the codemod then 😉
